core/src/main/kotlin/org/rewedigital/dialog/model/google/UserStorageHolder.kt

```diff
@@ -0,0 +1,74 @@
+package org.rewedigital.dialog.model.google
+
+import java.util.*
+
+interface UserStorageHolder {
+    var userStorage: String?
+}
+
+val UserStorageHolder.userData: DataStorage
+get() = DataStorage(this)
+
+class DataStorage(private val userData : UserStorageHolder) : LinkedHashMap<String, String>() {
+    init {
+        val data = userData.userStorage
+        data?.let {
+            val dataPos = data.indexOf("data")
+            val firstBracket = data.indexOf("{")
+            val secondBracket = data.indexOf("{", 1)
+            val seemsValid =
+                firstBracket == 0 && dataPos > firstBracket && secondBracket > dataPos && data.endsWith("}}")
+            if (seemsValid) {
+                // remove known pre- and suffix
+                data.substring(secondBracket + 1, data.length - 2)
+                    .split(',')
+                    .forEach { keyValuePair ->
+                        val (key, value) = keyValuePair.split(":")
+                        put(key.stripQuotesAndSlashes(), value.stripQuotesAndSlashes())
+                    }
+            }
+        }
+    }
+
+    private fun String.stripQuotesAndSlashes(): String {
+        return trim().trim('"').replace("\\\\", "\\")
+    }
+
+    override fun clear() {
+        super.clear()
+        userData.userStorage = asJson()
+    }
+
+    override fun putAll(from: Map<out String, String>) {
+        super.putAll(from)
+        userData.userStorage = asJson()
+    }
+
+    override fun put(key: String, value: String) =
+        super.put(key, value).also {
+            userData.userStorage = asJson()
+        }
+
+    override fun remove(key: String) =
+        super.remove(key).also {
+            userData.userStorage = asJson()
+        }
+
+    override fun remove(key: String, value: String)=
+        super.remove(key, value).also {
+            userData.userStorage = asJson()
+    }
+
+    val userId: String
```

Comment:
For migration it would be nice to look if there is still an userId. You could save the current userId from the user object to the userdata and if there is no userId you can create one like you already do.

Reply:
Good point, I will check how I can achieve that